cmd/zoekt-sourcegraph-indexserver/merge.go

```diff
@@ -0,0 +1,237 @@
+		maxSizeBytes:    int64(maxSize) * 1024 * 1024,
+		cutoffDate:      time.Now().AddDate(0, 0, -days),
+	}
+	compounds, excluded := generateCompounds(shards, opts)
+
+	debug.Printf("generated %d compounds and %d excluded repositories\n", len(compounds), len(excluded))
+	if len(compounds) == 0 {
+		return nil
+	}
+
+	var totalSizeBytes int64 = 0
+	totalShards := 0
+	for ix, comp := range compounds {
+		debug.Printf("compound %d: merging %d shards with total size %.2f MiB\n", ix, len(comp.shards), float64(comp.size)/(1024*1024))
+		if !simulate {
+			err := callMerge(comp.shards)
+			if err != nil {
+				return err
+			}
+			moveAll(backupDir, comp.shards)
+		}
+		totalShards += len(comp.shards)
+		totalSizeBytes += comp.size
+	}
+
+	debug.Printf("total size: %.2f MiB, number of shards merged: %d\n", float64(totalSizeBytes)/(1024*1024), totalShards)
+	return nil
+}
+
+// loadShards returns all simple shards in dir without .meta files.
+func loadShards(dir string) []shard {
+	d, err := os.Open(dir)
+	if err != nil {
+		debug.Printf("failed to loadShards: %s", dir)
+		return []shard{}
+	}
+	defer d.Close()
+	names, _ := d.Readdirnames(-1)
+
+	shards := make([]shard, 0, len(names))
+	for _, n := range names {
+		path := filepath.Join(dir, n)
+
+		fi, err := os.Stat(path)
+		if err != nil {
+			debug.Printf("stat failed for %s: %s", n, err)
+			continue
+		}
+
+		if fi.IsDir() || filepath.Ext(path) != ".zoekt" || strings.HasPrefix(filepath.Base(path), "compound-") {
+			continue
+		}
+
+		repos, _, err := zoekt.ReadMetadataPath(path)
+		if err != nil {
+			debug.Printf("failed to load metadata for %s\n", filepath.Base(path))
+			continue
+		}
+		if len(repos) != 1 {
+			debug.Printf("expected %s to be a simple shard, but encountered %d repos", n, len(repos))
+			continue
+		}
+
+		rank, err := strconv.ParseFloat(repos[0].RawConfig["rank"], 64)
+		if err != nil {
+			debug.Printf("error parsing rank %s for shard: %s: %s, setting rank to 0", repos[0].RawConfig["rank"], n, err)
+			rank = 0
+		}
+
+		shards = append(shards, shard{
+			Repo:      n,
+			Path:      path,
+			ModTime:   fi.ModTime(),
+			SizeBytes: fi.Size(),
+			Rank:      rank,
+		})
+	}
+	return shards
+}
+
+type compound struct {
+	shards []shard
+	size   int64
+}
+
+type compoundOpts struct {
+	targetSizeBytes int64
+	maxSizeBytes    int64
+	cutoffDate      time.Time
+}
+
+// generateCompounds encodes the merge policy:
+//   - merge shards that are older than opt.cutoffDate
+//   - merge shards that are smaller than opt.maxSizeBytes
+//   - merge shards with similar rank
+//   - compound shards should be larger than opt.targetSizeBytes
+func generateCompounds(shards []shard, opt compoundOpts) ([]compound, []shard) {
+	cur := 0
+	for ix, s := range shards {
+		if s.ModTime.After(opt.cutoffDate) || s.SizeBytes > opt.maxSizeBytes {
+			shards[cur], shards[ix] = shards[ix], shards[cur]
+			cur++
+			continue
+		}
+	}
+	if cur == len(shards) {
```

Comment:
nit: maybe check below that `shards` is empty. Easier to validate correct since you don't need to understand chat cur == len(shards) means.